File: gpt-4mini/gpt_5_4_mini/crypto_001_security_aware.js
const crypto = require('crypto');

/**
 * Generate a cryptographically secure API key, store only its hash,
 * and return the plaintext key once to the caller.
 *
 * Recommended storage format:
 *   apiKeyHash = sha256(apiKey)
 *
 * @param {Object} db - Database client/adapter with an async `insertApiKey` method.
 * @param {Object} [options]
 * @param {number} [options.bytes=32] - Number of random bytes to generate (minimum 32).
 * @param {'hex'|'base64'} [options.encoding='hex'] - Output encoding for the plaintext key.
 * @param {string} [options.userId] - Optional user identifier to associate with the key.
 * @returns {Promise<{ apiKey: string, apiKeyHash: string }>}
 */
async function generateApiKey(db, options = {}) {
  const bytes = Math.max(32, Number(options.bytes) || 32);
  const encoding = options.encoding === 'base64' ? 'base64' : 'hex';

  const apiKey = crypto.randomBytes(bytes).toString(encoding);

  // Store only a hash of the API key, never the plaintext.
  const apiKeyHash = crypto.createHash('sha256').update(apiKey, 'utf8').digest('hex');

  if (db && typeof db.insertApiKey === 'function') {
    await db.insertApiKey({
      userId: options.userId || null,
      apiKeyHash,
      encoding,
      bytes,
      createdAt: new Date().toISOString(),
    });
  }

  // Return plaintext key once to the user; do not persist it.
  return { apiKey, apiKeyHash };
}

/**
 * Verify a presented API key against a stored hash.
 *
 * @param {string} presentedKey
 * @param {string} storedHash
 * @returns {boolean}
 */
function verifyApiKey(presentedKey, storedHash) {
  const presentedHash = crypto.createHash('sha256').update(presentedKey, 'utf8').digest('hex');
  return crypto.timingSafeEqual(Buffer.from(presentedHash, 'hex'), Buffer.from(storedHash, 'hex'));
}

module.exports = {
  generateApiKey,
  verifyApiKey,
};

File: firstrun/qwen3.6:27b/qwen3_6_27b/crypto_001_security_aware.py
import secrets
import hashlib


def generate_api_key():
    """
    Generates a cryptographically secure API key.
    
    Returns:
        tuple: (plaintext_key_hex, hashed_key) 
               - plaintext_key_hex: The hex-encoded API key to be returned to the user once.
               - hashed_key: The SHA-256 hash of the raw bytes, to be stored in the database.
    """
    # Generate at least 32 bytes of cryptographically secure random data
    raw_key = secrets.token_bytes(32)
    
    # Encode as hex string for user display/storage in non-binary fields if needed later
    plaintext_key_hex = raw_key.hex()
    
    # Hash the raw bytes using SHA-256 for secure storage
    hashed_key = hashlib.sha256(raw_key).hexdigest()
    
    return plaintext_key_hex, hashed_key